Ensure ObservedWatch.path is a string (fix #651)

rec opened the pull request.

BoboTiG commented:
The fix is breaking too many tests.
Perhaps could you try something like
```python
    if isinstance(path, os.PathLike):
        self._path = os.fspath(path)
```
But that will work for Python 3.6+ only.

rec commented:
When I submitted this, the tests on master were also broken!  :-/

I can easily update this now the tests are working...

rec commented:
This does need to get fixed, because that trap is annoying to debug if you run into it.

rec commented:
Sorry for all the traffic! I have two working branches.

[One of them](https://github.com/rec/watchdog/commit/d2280d481e3fbf528b705eb13ef7c77ab29a43f5) uses your suggestion.  The tests pass for all Pythons, but it only fixes the issue for Python 3.6 and up.

[The other one](https://github.com/rec/watchdog/commit/132159252994e831015540ee108f5a49aa1e0592) uses a slightly different mechanism.  The tests pass, and it fixes the issue for all Python 3.

I pulled the second one into this review.

BoboTiG commented:
OK, go for the second solution. Could you add a test and the NEWs entry in the changelog then?

rec commented:
All right, I changed the changelog, and added a somewhat lame test that at least covers the codepath.

Unfortunately, adding a full test that triggers the bad behavior can't be done, because it causes a TRAP.

All the tests passed, and I rebased the fixup commit in!

BoboTiG commented on changelog.rst:
```diff
@@ -8,6 +8,7 @@ Changelog
 
 2020-0x-xx • `full history <https://github.com/gorakhargosh/watchdog/compare/v0.10.2...master>`__
 
+- Ensure ObservedWatch.path is a string (`#655 <https://github.com/gorakhargosh/watchdog/pull/651>`_)
```
```suggestion
- Ensure ``ObservedWatch.path`` is a string (`#651 <https://github.com/gorakhargosh/watchdog/pull/651>`_)
```
Let's wait for the tests to pass first. Then you can amend the commit and I will merge right after.

BoboTiG commented on changelog.rst:
```diff
@@ -8,6 +8,7 @@ Changelog
 
+- Ensure ObservedWatch.path is a string (`#655 <https://github.com/gorakhargosh/watchdog/pull/651>`_)
 - [inotify] Allow to monitor single file (`#655 <https://github.com/gorakhargosh/watchdog/pull/655>`__)
 - Thanks to our beloved contributors: @brant-ruan
```
```suggestion
- Thanks to our beloved contributors: @brant-ruan, @rec
```

BoboTiG commented:
Thanks for the patch @rec :)

rec commented:
Not at all, thanks for your excellent library, which I would have had to
write if it didn't already exist - and probably not done such a good job
of.  :-D

I make good use of this in my upcoming library backer
<https://github.com/rec/backer> (from my series the Coroniad
<https://medium.com/@TomSwirly/the-coroniad-8b7f7c0c82f4>).

Have a great day, and stay safe!

On Tue, Apr 28, 2020 at 3:01 PM Mickaël Schoentgen <notifications@github.com>
wrote:

> Thanks for the patch @rec <https://github.com/rec> :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gorakhargosh/watchdog/pull/652#issuecomment-620591361>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB53MVEIR5BMFBDOXNVPRTRO3HRBANCNFSM4K4KNABA>
> .
>


-- 
     /t

PGP Key: https://flowcrypt.com/pub/tom.ritchford@gmail.com
*https://tom.ritchford.com <https://tom.ritchford.com>*
*https://tom.swirly.com <https://tom.swirly.com>*
